clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java

```diff
@@ -0,0 +1,359 @@
+        : DEFAULT_REFRESH_INTERVAL_IN_SECONDS;
+    this.transportClient = transportClient;
+    this.d2ServiceDiscovery = new D2ServiceDiscovery();
+    this.clusterDiscoveryD2ServiceName = transportClient.getServiceName();
+    this.compressorFactory = new CompressorFactory();
+    this.clusterStats = clientConfig.getClusterStats();
+  }
+
+  @Override
+  public int getCurrentStoreVersion() {
+    return currentVersion.get();
+  }
+
+  @Override
+  public int getPartitionId(int version, ByteBuffer key) {
+    PartitionerPair partitionerPair = versionPartitionerMap.get(version);
+    if (partitionerPair == null) {
+      throw new VeniceClientException("Unknown version number: " + version + " for store: " + storeName);
+    }
+    return partitionerPair.getVenicePartitioner().getPartitionId(key, partitionerPair.getPartitionCount());
+  }
+
+  @Override
+  public List<String> getReplicas(int version, int partitionId) {
+    String key = getVersionPartitionMapKey(version, partitionId);
+    return readyToServeInstancesMap.getOrDefault(key, Collections.emptyList());
+  }
+
+  private String getVersionPartitionMapKey(int version, int partition) {
+    return version + VERSION_PARTITION_SEPARATOR + partition;
+  }
+
+  /**
+   * Given a version partitioner key formatted as "1_0" which represents partition 0 of version 1, return the version
+   * number.
+   * @param key
+   * @return version number
+   */
+  private int getVersionFromKey(String key) {
+    return Integer.parseInt(key.split(VERSION_PARTITION_SEPARATOR)[0]);
+  }
+
+  @Override
+  public void start() {
+    // perform cluster discovery work upfront and retrieve the server d2 service name
+    discoverD2Service(false);
+
+    // build a base for future metadata updates then start periodic refresh
+    refresh();
+  }
+
+  private void discoverD2Service(boolean retryOnFailure) {
+    if (isServiceDiscovered) {
+      return;
+    }
+    synchronized (this) {
+      if (isServiceDiscovered) {
+        return;
+      }
+      transportClient.setServiceName(clusterDiscoveryD2ServiceName);
+      String serverD2ServiceName =
+          d2ServiceDiscovery.find(transportClient, storeName, retryOnFailure).getServerD2Service();
+      transportClient.setServiceName(serverD2ServiceName);
+      isServiceDiscovered = true;
+    }
+  }
+
+  /**
+   * Update is only performed if the version from the fetched metadata is different from the local version. We evict
+   * old values as we perform updates, while making sure we keep the two most recent versions in the cache.
+   * @param onDemandRefresh
+   */
+  private synchronized void updateCache(boolean onDemandRefresh) throws InterruptedException {
+    // call the METADATA endpoint
+    try {
+      byte[] body = fetchMetadata().get().getBody();
+      RecordDeserializer<MetadataResponseRecord> metadataResponseRecordRecordDeserializer =
+          SerializerDeserializerFactory.getAvroGenericDeserializer(MetadataResponseRecord.SCHEMA$);
```

Comment:
Can you see if we can use `SpecifcRecord` for this? Using SR will be more efficient and you won't have to extract each and every field manually.

Also, please check if we can remove `PartitionerPair` after we move to `SpecificRecord`

Reply:
Also, let's use `FastSerializerDeserializerFactory`

Reply:
Switched to `SpecifcRecord` and removed the `PartitionerPair`. Split the map into two maps: one that stores version -> `VenicePartitioner` and another that stores version -> partition count.